packages/salesforce-adapter/src/filters/workflow_rule.ts

```diff
@@ -0,0 +1,161 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import _ from 'lodash'
+import { collections } from '@salto-io/lowerdash'
+import { logger } from '@salto-io/logging'
+import {
+  Element, ReferenceExpression, ElemID, Value, Values, isReferenceExpression,
+} from '@salto-io/adapter-api'
+import {
+  findInstances,
+} from '@salto-io/adapter-utils'
+import { FilterCreator } from '../filter'
+import { parentApiName } from './utils'
+import { groupByAPIName, ApiNameMapping } from './instance_references'
+import {
+  SALESFORCE, API_NAME_SEPERATOR,
+  WORKFLOW_ACTION_ALERT_METADATA_TYPE,
+  WORKFLOW_FIELD_UPDATE_METADATA_TYPE,
+  WORKFLOW_FLOW_ACTION_METADATA_TYPE,
+  WORKFLOW_OUTBOUND_MESSAGE_METADATA_TYPE,
+  WORKFLOW_RULE_METADATA_TYPE,
+  WORKFLOW_TASK_METADATA_TYPE,
+} from '../constants'
+
+const log = logger(module)
+const { makeArray } = collections.array
+
+export const WORKFLOW_RULE_TYPE_ID = new ElemID(
+  SALESFORCE,
+  WORKFLOW_RULE_METADATA_TYPE,
+)
+
+export type WorkflowActionType = 'Alert' | 'FieldUpdate' | 'FlowAction' | 'OutboundMessage' | 'Task'
+export const WORKFLOW_ACTION_TYPE_MAPPING: Record<WorkflowActionType, string> = {
+  Alert: WORKFLOW_ACTION_ALERT_METADATA_TYPE,
+  FieldUpdate: WORKFLOW_FIELD_UPDATE_METADATA_TYPE,
+  FlowAction: WORKFLOW_FLOW_ACTION_METADATA_TYPE,
+  OutboundMessage: WORKFLOW_OUTBOUND_MESSAGE_METADATA_TYPE,
+  Task: WORKFLOW_TASK_METADATA_TYPE,
+}
+
+
+export type WorkflowActionReference = {
+  name: string | ReferenceExpression
+  type: WorkflowActionType
+}
+
+type WorkflowRule = Values & {
+  actions: WorkflowActionReference | WorkflowActionReference[]
+}
+
+/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
+function isWorkflowActionReference(value: any): value is WorkflowActionReference {
+  return value !== undefined && (
+    (_.isString(value.name) || isReferenceExpression(value.name))
+    && Object.keys(WORKFLOW_ACTION_TYPE_MAPPING).includes(value.type)
+  )
+}
+
+/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
+function isWOrkflowRule(value: any): value is WorkflowRule {
+  return value !== undefined && (
+    (_.isArray(value.actions) && value.actions.every(isWorkflowActionReference))
+    || isWorkflowActionReference(value.actions)
+  )
+}
+
+/**
+ * Replace the action name with a reference, based on the parent and action type.
+ *
+ * @param action            The action to update
+ * @param parentObjName     The name of the containing workflow rule's parent object
+ * @param apiNameToElemIDs  Known element ids, mapped by API name and metadata type
+ */
+const convertActionToReference = (
+  action: WorkflowActionReference,
+  parentObjName: string,
+  apiNameToElemIDs: ApiNameMapping,
+  workflowRuleName: string,
+): void => {
+  if (action === undefined || !_.isString(action.name)) {
+    return
+  }
+
+  const findReference = (name: string): Value => {
```

Comment:
nit: return type can be more specific
```suggestion
  const findReference = (name: string): ReferenceExpression | undefined => {
```